File: src/main/java/at/fhlinz/imagej/Calibration.java
/****************************************************************************
 *
 * MIT License
 *
 * Copyright (C) 2020 Fabian Hauser
 *
 * Author: Fabian Hauser <fabian.hauser@fh-linz.at>
 * University of Applied Sciences Upper Austria - Linz - Austria
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 ****************************************************************************/

package at.fhlinz.imagej;

import ij.gui.Plot;
import ij.gui.PlotWindow;
import java.awt.Color;
import java.io.File;
import java.io.FileInputStream;
import java.util.HashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.apache.commons.math3.analysis.UnivariateFunction;
import org.apache.commons.math3.analysis.polynomials.PolynomialFunction;
import org.apache.commons.math3.analysis.polynomials.PolynomialSplineFunction;
import org.apache.commons.math3.linear.Array2DRowRealMatrix;
import org.apache.commons.math3.linear.ArrayRealVector;
import org.apache.commons.math3.linear.LUDecomposition;
import org.apache.commons.math3.linear.RealMatrix;
import org.apache.commons.math3.linear.RealVector;
import org.apache.commons.math3.optim.MaxEval;
import org.apache.commons.math3.optim.nonlinear.scalar.GoalType;
import org.apache.commons.math3.optim.univariate.BrentOptimizer;
import org.apache.commons.math3.optim.univariate.SearchInterval;
import org.apache.commons.math3.optim.univariate.UnivariateObjectiveFunction;
import org.apache.commons.math3.optim.univariate.UnivariatePointValuePair;

/**
 * Calibration class to handle astigmatism single molecule localization 
 * microscopy calibration files using cubic b-splines
 * 
 * Input is a YAML file:
 * !!at.calibration.bspline { 
 *    theta: -0.103174,
 *    pixelSize: 166.667, 
 *    knot0x: 0.986468, knot0y: 0.402337, knot0z: 0, 
 *    ...
 *    knot9x: 0.3007, knot9y: 0.558788, knot9z: 2000
 * }
 * 
 * Where theta is the rotation angle of the elliptical Gaussian functions in 
 * RAD, pixelSize of in nm, knot0 .. knotN where N is the last spline knot,
 * knotX & knotY in pixel and knotZ in nm
 * 
 * @author Fabian Hauser
 */
public class Calibration {
    private final HashMap<String, Double> _parameters;
    private double _pixelSizeUM;
    private double _theta;
    private PolynomialSplineFunction _wxSpline;
    private PolynomialSplineFunction _wySpline;
    private UnivariateFunction _dwx;
    private UnivariateFunction _dwy;
    private double _minZ;
    private double _maxZ;
    private double _focusPlane;
    
    /**
     * Constructor
     */
    public Calibration() {
        _parameters = new HashMap<String, Double>();
        _pixelSizeUM = 1.0;
        _theta = 0.0;
        _wxSpline = null;
        _wySpline = null;
    }
            
    /**
     * Load a calibration yaml file. Supported are cubic bspline representation 
     * (at.calibration.bspline)
     * Needed parameters are 'pixelSize' in nm and 'angle'/'theta'
     * @param fileName YAML file with parameter
     * @return returns true if the YAML file could be loaded
     */
    public boolean load(String fileName)
    {
        try {
            File file = new File(fileName);   
            FileInputStream stream = new FileInputStream(file);
            byte [] bytes = new byte[(int)file.length()];
            if (stream.read(bytes) != file.length())
                return false;
            String data = new String(bytes, "UTF8");
            
            Matcher matcher = Pattern.compile("!!([\\w\\.]+)\\s*").matcher(data);
            if (!matcher.find())
                return false;
            
            String typeString = matcher.group(1);
            if (!typeString.equals("at.calibration.bspline")) {
                System.err.printf("LookUpSTORM: Calibration: Unknown type '%s'!\n", typeString);
                return false;
            }
            
            _parameters.clear();
            
            // parse parameters
            Pattern p = Pattern.compile("(\\w*):\\s*([\\-]*[0-9+][.\\w]*[-\\w]*)");
            matcher = p.matcher(data.substring(matcher.end()));
            while (matcher.find()) {
                _parameters.put(matcher.group(1), Double.parseDouble(matcher.group(2)));
            }
            if (_parameters.isEmpty())
                return false;
            
            // find important parameters
            if (_parameters.containsKey("pixelSize"))
                _pixelSizeUM = _parameters.get("pixelSize") / 1000;
            
            if (_parameters.containsKey("theta"))
                _theta = _parameters.get("theta");
            else if (_parameters.containsKey("angle"))
                _theta = _parameters.get("angle");
            
            generateSplines();
            
            // determine focal plane
            if (_parameters.containsKey("focalPlane"))
                _focusPlane = _parameters.get("focalPlane");
            else
                _focusPlane = calculateFocalPlane();
            
            return true;
        } catch(Exception e) {
            return false;
        }
    }
    
    /**
     * @return the focus plane in nm (axial position)
     */
    public double getFocusPlane() {
        return _focusPlane;
    }
    
    /**
     * @return get axial range in nm
     */
    public double getAxialRange() {
        return (_maxZ - _minZ);
    }
    
    /**
     * @return the minimum of the z position in nm
     */
    public double getMinZ() {
        return (_maxZ - _minZ);
    }
    
    /**
     * @return the maximum of the z position in nm
     */
    public double getMaxZ() {
        return (_maxZ - _minZ);
    }
    
    /**
     * @return the rotation theta in radians of the 2D elliptical Gaussian 
     * function
     */
    public double getTheta() {
        return _theta;
    }
    
    /**
     * @param z axial position in nm
     * @return a double array containing sigmaX & sigmaY in pixels from the 
     * provided axial position (in nm)
     */
    public double[] value(double z) {
        return new double[] {
            _wxSpline.value(z),
            _wySpline.value(z)
        };
    }
    
    /**
     * Returns an array with the values (sigmaX, sigmaY) and their derivative
     * in an array
     * @param z axial position in nm
     * @return array of [sigmaX,sigmaY,dSigmaY, dSigmaY]
     */
    public double[] valDer(double z) {
        return new double[] {
            _wxSpline.value(z),
            _wySpline.value(z),
            _dwx.value(z),
            _dwy.value(z)
        };
    }
    
    /**
     * Show the sigmaX & sigmaY curves of the loaded calibration
     */
    public void plot() {
        int n = 100;
        double[] wx = new double[n];
        double[] wy = new double[n];
        double[] x = new double[n];
        for (int i = 0; i < n; i++) {
            final Double z = _minZ + i * (_maxZ-_minZ) / (n-1);
            x[i] = z;
            wx[i] = _wxSpline.value(z);
            wy[i] = _wySpline.value(z);
        }
        Plot p = new Plot("Calibration", "z position / nm", "width / pixels");
        p.setLineWidth(1);

        p.setColor(new Color(255, 0, 0));
        p.addPoints(x, wx, PlotWindow.LINE);

        p.setColor(new Color(0, 0, 255));
        p.addPoints(x, wy, PlotWindow.LINE);

        p.addLegend("SigmaX\nSigmaY");
        p.show();
    }
    
    /** 
     * Calculate cubic bspline polynomials from array values
     * @param x x-values
     * @param y y-values
     * @return PolynomialSplineFunction containing the bspline polynomials
     */
    private PolynomialSplineFunction generateSpline(double[] x, double[] y) {
        final int N = x.length;
        final double h = x[1] - x[0];        
        double [][] dA = new double[N-2][N-2];
        
        RealMatrix A = new Array2DRowRealMatrix(dA, false);
        RealVector b = new ArrayRealVector(N-2);
        
        // calculate matrix for cubic bspline polynomal
        for (int j = 0; j < N-2; j++) {
            b.setEntry(j, (y[j] - 2*y[j+1] + y[j+2]) * 6 / (h*h));
            dA[j][j] = 4;
            if (j > 0) dA[j][j-1] = 1;
            if (j < N-3) dA[j][j+1] = 1;
        }
        LUDecomposition lu = new LUDecomposition(A);
        RealVector lux = lu.getSolver().solve(b);
        
        double[] m = new double[N];
        for (int i = 0; i < lux.getDimension(); i++)
            m[i+1] = lux.getEntry(i);
        
        PolynomialFunction[] polynomials = new PolynomialFunction[N - 1];
        for (int i = 0; i < polynomials.length; i++) {
            double[] p = new double[4];
            p[3] = (m[i+1] - m[i]) / (6.0 * h);
            p[2] = 0.5 * m[i];
            p[1] = (y[i+1] -  y[i]) / h - (m[i+1] + 2.0 * m[i]) * h / 6.0;
            p[0] = y[i];
            polynomials[i] = new PolynomialFunction(p);
        }
        return new PolynomialSplineFunction(x, polynomials);
    }
    
    private void generateSplines() {
        int knots = 0;
        while(_parameters.containsKey("knot"+knots+"x")) {
            knots++;
        }
        double[] xKnots = new double[knots];
        double[] yKnots = new double[knots];
        double[] zKnots = new double[knots];
        for (int i = 0; i < knots; ++i) {
            xKnots[i] = _parameters.get("knot"+i+"x") / _pixelSizeUM;
            yKnots[i] = _parameters.get("knot"+i+"y") / _pixelSizeUM;
            zKnots[i] = _parameters.get("knot"+i+"z");
        }
        
        _minZ = zKnots[0];
        _maxZ = zKnots[zKnots.length-1];
        
        _wxSpline = generateSpline(zKnots, xKnots);
        _wySpline = generateSpline(zKnots, yKnots);
        
        _dwx = _wxSpline.derivative();
        _dwy = _wySpline.derivative();
    }
    
    private class FocusFunction implements UnivariateFunction {
        private final Calibration _cali;
        
        public FocusFunction(Calibration cali)
        {
            _cali = cali;
        }
        
        @Override
        public double value(double d) {
            double[] v = _cali.value(d);
            return Math.abs(v[0] - v[1]);
        }
    };
    
    private double calculateFocalPlane() {
        final double relativeTolerance = 1.49e-08;
        final double absoluteTolerance = 1.49e-08;
        FocusFunction fn = new FocusFunction(this);
        final BrentOptimizer optimizer = new BrentOptimizer(relativeTolerance, absoluteTolerance);
        UnivariatePointValuePair p = optimizer.optimize(
                new UnivariateObjectiveFunction(fn), 
                new SearchInterval(_minZ, _maxZ, (_maxZ-_minZ)/2), 
                new MaxEval(100),
                GoalType.MINIMIZE);
        return p.getPoint();
    }
    
}

File: LookUpSTORM_CPPDLL/src/Calibration.cpp
/****************************************************************************
 *
 * MIT License
 *
 * Copyright (C) 2021 Fabian Hauser
 *
 * Author: Fabian Hauser <fabian.hauser@fh-linz.at>
 * University of Applied Sciences Upper Austria - Linz - Austria
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 ****************************************************************************/

#include "Calibration.h"

#include "LinearMath.h"
#include "brent.hpp"

#include <vector>
#include <iostream>
#include <fstream>
#include <regex>

namespace LookUpSTORM
{

struct Knot
{
    double x;
    double y;
    double z;

    inline constexpr const double &operator[](size_t dir) const 
    { return dir == 1 ? y : x; }
};



class FocusFunction : public brent::func_base {
    const Calibration* m_cali;
public:
    inline FocusFunction(const Calibration* c) : m_cali(c) {}
    inline virtual double operator() (double z)
    {
        const auto v = m_cali->value(z);
        return std::abs(v.first - v.second);
    }
};

class CalibrationPrivate
{
public:
    inline CalibrationPrivate()
        : h(0.0)
        , theta(0.0)
        , pixelSize(1.0)
        , focalPlane(0.0)
    {}

    std::vector<Knot> knots;
    double h;
    Matrix coeffs[2];
    double theta;
    double pixelSize;
    double focalPlane;
    Parameters parameters;

    inline bool parseParameters();
    inline bool generateSplines();
};

inline bool CalibrationPrivate::parseParameters()
{
    pixelSize = (parameters.count("pixelSize") > 0 ? parameters["pixelSize"] / 1000.0 : 1.0);

    int knotNr = 0;
    std::string knotName = "knot" + std::to_string(knotNr);
    while (parameters.count(knotName + 'x') > 0) {
        if (parameters.count(knotName + 'y') < 1 || parameters.count(knotName + 'z') < 1)
            break;
        knots.push_back({ parameters[knotName + 'x'] / pixelSize, 
                          parameters[knotName + 'y'] / pixelSize, 
                          parameters[knotName + 'z'] });
        ++knotNr;
        knotName = "knot" + std::to_string(knotNr);
    }
    if (knots.empty()) {
        std::cout << "Calibration: Cannot import calibration because it doesn't contains knots!" << std::endl;
        return false;
    }

    focalPlane = (knots.back().z - knots.front().z) * 0.5;

    if (parameters.count("angle") > 0) {
        theta = parameters["angle"];
    }
    else if (parameters.count("theta") > 0) {
        theta = parameters["theta"];
    }
    else {
        std::cout << "Calibration: The angle theta is not defined!" << std::endl;
        return false;
    }
    return true;
}

inline bool CalibrationPrivate::generateSplines()
{
    // solve equation for cubic b-splines from knots
    const size_t N = knots.size();
    Vector b(N - 2);
    Matrix A(N - 2, N - 2);
    std::vector<int> ipiv(N - 2);

    std::vector<double> m(N);

    h = (knots[1].z - knots[0].z);

    for (size_t dir = 0; dir < 2; ++dir) {
        A.setZero();
        for (size_t j = 0; j < b.size(); ++j) {
            b[j] = (knots[j][dir] - 2.0 * knots[j + 1][dir] + knots[j + 2][dir]) * 6.0 / (h * h);
            A(j, j) = 4.0;
            if (j > 0) A(j, j - 1) = 1.0;
            if (j < N - 3) A(j, j + 1) = 1.0;
        }
        // try to solve the equation system
        if (LAPACKE::dgesv(A, ipiv.data(), b) != 0)
            return false;

        // prepare coefficent matrix
        m.front() = 0.0;
        m.back() = 0.0;
        for (size_t i = 0; i < b.size(); i++)
            m[i + 1] = b[i];

        coeffs[dir] = Matrix(N - 1, 4);
        for (size_t i = 0; i < N - 1; ++i) {
            coeffs[dir](i, 3) = knots[i][dir];
            coeffs[dir](i, 2) = (knots[i + 1][dir] - knots[i][dir]) / h
                - (m[i + 1] + 2 * m[i]) * h / 6.0;
            coeffs[dir](i, 1) = m[i] / 2;
            coeffs[dir](i, 0) = (m[i + 1] - m[i]) / (6 * h);
        }
    }
    return true;
}

} // namespace LookUpSTORM

using namespace LookUpSTORM;

Calibration::Calibration()
    : d(new CalibrationPrivate)
{
}

LookUpSTORM::Calibration::~Calibration()
{
    delete d;
}

bool Calibration::load(const std::string& fileName)
{
    std::string line;
    std::ifstream file;
    file.open(fileName, std::ifstream::in);
    if (!file) {
        std::cerr << "Calibration: Could not open calibration file! (" << fileName << ")" << std::endl;
        return false;
    }

    std::string data;
    while (std::getline(file, line))
        data += line;

    file.close();

    return parseJAML(data);
}

bool LookUpSTORM::Calibration::parseJAML(const std::string& data)
{
    if (data.empty()) {
        std::cerr << "Calibration: Calibration file is empty!" << std::endl;
        return false;
    }

    d->knots.clear();
    d->parameters.clear();

    try {
        std::regex re("!!([\\w\\.]+)\\s*");
        std::smatch match;
        if (std::regex_search(data, match, re) && match.size() > 1) {
            const std::string typeName = match.str(1);

        }
        else {
            std::cerr << "Calibration: Invalid header!" << std::endl;
            return false;
        }

        std::regex param("(\\w*):\\s*([\\-]*[0-9+][.\\w]*[-\\w]*)");
        std::sregex_iterator next(data.begin(), data.end(), param);
        std::sregex_iterator end;
        while (next != end) {
            std::smatch match = *next;
            next++;
            if (match.size() != 3) continue;
            std::string key = match.str(1);
            const double value = std::stod(match.str(2));
            d->parameters.insert({ key, value });
        }
    }
    catch (std::regex_error& e) {
        // Syntax error in the regular expression
        std::cerr << "Calibration: Regex error: " << e.what() << std::endl;
        return false;
    }

    if (!d->parseParameters() || !d->generateSplines())
        return false;

    if (d->parameters.count("focalPlane") > 0) {
        d->focalPlane = d->parameters["focalPlane"];
    }
    else {
        // calculate focal plane
        FocusFunction func(this);
        const double e = std::sqrt(brent::r8_epsilon());
        const double t = std::sqrt(brent::r8_epsilon());

        const double min = d->knots.front().z;
        const double max = d->knots.back().z;
        brent::glomin(min, max, (max - min) / 2, 10, e, t, func, d->focalPlane);
    }

    return true;
}

bool Calibration::generateSpline()
{
    return false;
}

std::pair<double, double> Calibration::value(double z) const
{
    std::pair<double, double> ret{ 0.0, 0.0 };
    const size_t i = std::min(std::max(0ull, static_cast<size_t>(std::floor((z - d->knots[0].z) / d->h))), d->knots.size() - 2);
    const double dx = (z - d->knots[i].z);
    for (size_t term = 0; term < 4; ++term) {
        ret.first = ret.first * dx + d->coeffs[0](i, term);
        ret.second = ret.second * dx + d->coeffs[1]( i, term);
    }
    return ret;
}

std::pair<double, double> Calibration::dvalue(double z) const
{
    std::pair<double, double> ret{ 0.0, 0.0 };
    const size_t i = std::min(std::max(0ull, static_cast<size_t>(std::floor((z - d->knots[0].z) / d->h))), d->knots.size() - 2);
    const double dx = (z - d->knots[i].z);
    for (size_t term = 0; term < 3; ++term) {
        ret.first = ret.first * dx + (3 - term) * d->coeffs[0](i, term);
        ret.second = ret.second * dx + (3 - term) * d->coeffs[1](i, term);
    }
    return ret;
}

std::tuple<double, double, double, double> Calibration::valDer(double z) const
{
    std::tuple<double, double, double, double> ret{ 0.0, 0.0, 0.0, 0.0 };
    const size_t i = std::min(std::max(0ull, static_cast<size_t>(std::floor((z - d->knots[0].z) / d->h))), d->knots.size() - 2);
    const double dx = (z - d->knots[i].z);
    for (size_t term = 0; term < 3; ++term) {
        std::get<0>(ret) = std::get<0>(ret) * dx + d->coeffs[0](i, term);
        std::get<1>(ret) = std::get<1>(ret) * dx + d->coeffs[1](i, term);
        std::get<2>(ret) = std::get<2>(ret) * dx + (3 - term) * d->coeffs[0](i, term);
        std::get<3>(ret) = std::get<3>(ret) * dx + (3 - term) * d->coeffs[1](i, term);
    }
    std::get<0>(ret) = std::get<0>(ret) * dx + d->coeffs[0](i, 3);
    std::get<1>(ret) = std::get<1>(ret) * dx + d->coeffs[1](i, 3);
    return ret;
}

size_t LookUpSTORM::Calibration::knots() const
{
    return d->knots.size();
}

std::tuple<double, double, double> LookUpSTORM::Calibration::knot(size_t i) const
{
    return { d->knots[i].x, d->knots[i].y, d->knots[i].z };
}

double Calibration::theta() const
{
    return d->theta;
}

double LookUpSTORM::Calibration::pixelSize() const
{
    return d->pixelSize;
}

double Calibration::focalPlane() const
{
    return d->focalPlane;
}

double Calibration::minZ() const
{
    return d->knots.empty() ? 0.0 : d->knots.front().z;
}

double Calibration::maxZ() const
{
    return d->knots.empty() ? 0.0 : d->knots.back().z;
}

const Parameters& LookUpSTORM::Calibration::parameters() const
{
    return d->parameters;
}
